api/workload/v2/workload.go

```diff
@@ -0,0 +1,219 @@
+	// that the workload should trust, keyed by the SPIFFE ID of the foreign domain.
+	// Unimplemented.
+	FederatedBundles map[string]*x509.CertPool
+}
+
+// Default returns the default SVID (the first in the list).
+//
+// See the SPIFFE Workload API standard Section 5.3
+// (https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Workload_API.md#53-default-identity)
+func (x *X509SVIDs) Default() *X509SVID {
+	return x.SVIDs[0]
+}
+
+// SVID is an X.509 SPIFFE Verifiable Identity Document.
+//
+// See https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md
+type X509SVID struct {
+	SPIFFEID     string
+	PrivateKey   crypto.Signer
+	Certificates []*x509.Certificate
+	TrustBundle  *x509.CertPool
+}
+
+// WorkloadX509SVIDWatcher is implemented by consumers who wish to be updated on SVID changes.
+type WorkloadX509SVIDWatcher interface {
+	// UpdateX509SVIDs indicates to the Watcher that the SVID has been updated
+	UpdateX509SVIDs(*X509SVIDs)
+
+	// OnError indicates an error occurred.
+	OnError(err error)
+
+	// OnConnection indicates a change in the connection state to the SPIFFE agent.
+	OnConnection(connected bool)
+}
+
+// X509SVIDClient interacts with the SPIFFE Workload API.
+type X509SVIDClient struct {
+	watcher  WorkloadX509SVIDWatcher
+	addr     string
+	wg       sync.WaitGroup
+	ctx      context.Context
+	cancelFn func()
+	backoff  *backoff
+}
+
+// Option configures the workload client.
+type Option func(*X509SVIDClient)
+
+// WithAddr specifies the unix socket address of the SPIFFE agent.
+func WithAddr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		addr:     DefaultAgentAddress,
+		watcher:  watcher,
+		ctx:      ctx,
+		cancelFn: cancel,
+		backoff:  newBackoff(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
+	c.wg.Add(1)
+	go c.run()
+	return nil
+}
+
+func (c *X509SVIDClient) run() {
+	defer c.wg.Done()
+
+	conn := c.newConn()
+	if conn == nil {
+		return
+	}
+	c.watcher.OnConnection(true)
+	defer c.watcher.OnConnection(false)
+	defer conn.Close()
+
+	c.watch(conn)
+}
+
+// establishes a new persistent connection, returns nil if a client can't be created within the context deadline
+func (c *X509SVIDClient) newConn() *grpc.ClientConn {
+	for {
+		conn, err := grpc.DialContext(c.ctx, c.addr, grpc.WithInsecure())
```

Comment:
that's good to know. and while it may be true today, it's not necessarily a contract of the DialContext API, so we still want to check for the error, right?

Reply:
We still want to check the error for sure. However, without WithBlock, this means that the "OnConnected(true)" will be misleading since the code won't now if the async dial has even succeeded by then. You don't really know that you've connected until FetchX509SVID has returned a stream.